contracts/deposit-contracts/contract/src/main.rs

```diff
@@ -0,0 +1,214 @@
+                .to_bytes()
+                .unwrap_or_revert(),
+        ),
+        SecurityBadge::Admin,
+    );
+    let admin_list: Option<Vec<Key>> =
+        get_optional_named_arg_with_user_errors(ADMIN_LIST, DepositError::InvalidAdminList);
+    if let Some(admin_list) = admin_list {
+        for admin in admin_list {
+            storage::dictionary_put(
+                security_badges_dict,
+                &base64::encode(admin.to_bytes().unwrap_or_revert()),
+                SecurityBadge::Admin,
+            );
+        }
+    };
+    let new_deposit_purse: URef = system::create_purse();
+    runtime::put_key(KAIROS_DEPOSIT_PURSE, new_deposit_purse.into());
+}
+
+#[no_mangle]
+pub extern "C" fn get_purse() {
+    let deposit_purse: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(ApiError::MissingKey)
+        .into_uref()
+        .unwrap_or_revert();
+    runtime::ret(
+        CLValue::from_t(deposit_purse.with_access_rights(AccessRights::ADD)).unwrap_or_revert(),
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn deposit() {
+    let temp_purse: URef = runtime::get_named_arg(RUNTIME_ARG_TEMP_PURSE);
+    let amount: U512 = runtime::get_named_arg(RUNTIME_ARG_AMOUNT);
+    let deposit_purse_uref: URef = runtime::get_key(KAIROS_DEPOSIT_PURSE)
+        .unwrap_or_revert_with(ApiError::MissingKey)
+        .into_uref()
+        .unwrap_or_revert_with(ApiError::UnexpectedKeyVariant);
+    system::transfer_from_purse_to_purse(temp_purse, deposit_purse_uref, amount, None)
+        .unwrap_or_revert();
+
+    let most_recent_deposit_counter_uref = runtime::get_key(KAIROS_MOST_RECENT_DEPOSIT_COUNTER)
+        .unwrap_or_revert_with(ApiError::MissingKey)
+        .into_uref()
+        .unwrap_or_revert();
+    let mut most_recent_deposit_counter_value: u64 =
+        storage::read(most_recent_deposit_counter_uref)
+            .unwrap_or_revert()
+            .unwrap_or_revert();
+    let new_deposit_record: Deposit = Deposit {
+        account: get_immediate_caller().unwrap_or_revert(),
+        amount,
+        timestamp: None,
+        processed: false,
+    };
+
+    let kairos_deposit_event_dict_uref = runtime::get_key(KAIROS_DEPOSIT_EVENT_DICT)
+        .unwrap_or_revert_with(ApiError::MissingKey)
+        .into_uref()
+        .unwrap_or_revert();
+    storage::dictionary_put::<Vec<u8>>(
+        kairos_deposit_event_dict_uref,
+        &most_recent_deposit_counter_value.to_string(),
+        bincode::serialize(&new_deposit_record).unwrap(),
+    );
+
+    most_recent_deposit_counter_value += 1u64;
+    storage::write(
+        most_recent_deposit_counter_uref,
+        most_recent_deposit_counter_value,
+    );
+}
+
+#[no_mangle]
+pub extern "C" fn incr_last_processed_deposit_counter() {
+    sec_check(vec![SecurityBadge::Admin]);
+    let last_processed_deposit_counter_uref =
+        runtime::get_key(KAIROS_LAST_PROCESSED_DEPOSIT_COUNTER)
+            .unwrap_or_revert_with(ApiError::MissingKey)
```

Comment:
With all these error messages we could probably improve the verbosity by adding what key is missing

Reply:
Yes that's something I can do. I actually thought about it before.